File: shafaet/lower_bound.cpp
#include<iostream>
using namespace std;

int nums[5] = {1, 2, 3, 4, 5};

int lower_bound(int n, int value) {
    int begin = 0, end = n - 1, mid, temp;
    while (begin <= end) {
        mid = (begin + end) / 2;
        temp = nums[mid];
        if (temp == value)
            end = mid - 1;
        else if (temp > value)
            end = mid - 1;
        else
            begin = mid + 1;
    }
    return begin;
}

int main() {
    cout << lower_bound(5, 0) << endl;
}

File: shafaet/lower_bound.py
def search_lower_bound(arr, value):
    begin = 0
    end = len(arr)
    while begin <= end:
        mid = (begin + end) // 2
        temp = arr[mid]
        if temp == value:
            end = mid - 1
        elif temp > value:
            end = mid - 1
        else:
            begin = mid + 1
    return begin


if __name__ == '__main__':
    arr = [1, 3, 3, 4, 5, 6]
    print(search_lower_bound(arr, 3))
